gui/mass-remove.lua

```diff
@@ -0,0 +1,336 @@
+
+MassRemoveUI.ATTRS {
+    action="remove_a",
+    marking=false,
+    mark=nil
+}
+
+--Helper functions.
+local function tablelength(T)
+  local count = 0
+  for _ in pairs(T) do count = count + 1 end
+  return count
+end
+
+--Needed because I couldn't find a direct way of getting a tile at certain coordinates.
+local function getTileAt(x, y, z)
+    local block = dfhack.maps.getTileBlock(x, y, z)
+    if block then
+        return block.designation[x%16][y%16]
+    end
+end
+
+--Helper to match a job of a particular type at tile (x,y,z) and run the callback function on the job.
+local function iterateJobs(jobType, x, y, z, callback)
+    local joblist = df.global.world.jobs.list.next
+
+    while joblist do
+        local job = joblist.item
+        joblist = joblist.next
+
+        if job.job_type == jobType and job.pos.x == x and job.pos.y == y and job.pos.z == z then
+            callback(job)
+        end
+    end
+end
+
+--Sorts and returns the given arguments.
+local function minToMax(...)
+    local args={...}
+    table.sort(args,function(a,b) return a < b end)
+    return table.unpack(args)
+end
+
+local function paintMapTile(dc, vp, cursor, pos, ...)
+    if not same_xyz(cursor, pos) then
+        local stile = vp:tileToScreen(pos)
+        if stile.z == 0 then
+            dc:map(true):seek(stile.x,stile.y):char(...):map(false)
+        end
+    end
+end
+
+function MassRemoveUI:init()
+    self.saved_mode = df.global.ui.main.mode
+    df.global.ui.main.mode=df.ui_sidebar_mode.LookAround
+end
+
+function MassRemoveUI:onDestroy()
+    persistTable.GlobalTable.massRemoveAction=self.action
+    df.global.ui.main.mode = self.saved_mode
+end
+
+function MassRemoveUI:changeSuspendState(x, y, z, new_state)
+    iterateJobs(df.job_type.ConstructBuilding, x, y, z, function(job) job.flags.suspend = new_state end)
+end
+
+function MassRemoveUI:suspend(x, y, z)
+    self:changeSuspendState(x, y, z, true)
+end
+
+function MassRemoveUI:unsuspend(x, y, z)
+    self:changeSuspendState(x, y, z, false)
+end
+
+function MassRemoveUI:removeConstruction(x, y, z)
+    dfhack.constructions.designateRemove(x, y, z)
+end
+
+--Construction removals can either be marked as dig on the tile itself, or picked up as jobs. This function checks both.
+function MassRemoveUI:unremoveConstruction(x, y, z)
+    local tile = getTileAt(x,y,z)
+    tile.dig = df.tile_dig_designation.No
+
+    iterateJobs(df.job_type.RemoveConstruction, x, y, z, function(job) dfhack.job.removeJob(job) end)
+end
+
+function MassRemoveUI:removeBuilding(x, y, z)
+    local building = dfhack.buildings.findAtTile(x, y, z)
+    if building then
+        dfhack.buildings.deconstruct(building)
+    end
+end
+
+function MassRemoveUI:unremoveBuilding(x, y, z)
+    local building = dfhack.buildings.findAtTile(x, y, z)
+    if building then
+        for i, job in ipairs(building.jobs) do
+            if job.job_type == df.job_type.DestroyBuilding then
+                building.jobs:erase(i)
```

Comment:
You probably also want to remove this from `world.jobs` too. I suspect that in this case, the job simply wouldn't be cancelled, but a DF crash can also result in some cases.

Actually, `dfhack.job.removeJob(job)` should remove the job from everywhere, so I'd recommend switching to that. Since you're starting with a reference to the building, I think the loop around here would still be needed.

Reply:
Good catch. I adopted this from `gui/advfort`, but that's an adventure mode plugin so it might be fine there. I did manage to use `dfhack.job.removeJob(job)` after I discovered it (it's not really in the docs). I used it in `MassRemoveUI:unremoveConstruction`, but that method was written after `MassRemoveUI:unremoveBuilding`, so I didn't go back to change it.

P.S. I've been testing this on a real fort I've been playing in because I'm dumb, so I sure hope those dangling jobs don't crash my fort down the line.

Fixed in ea34775b